035-ca-abstraction.md

```diff
@@ -0,0 +1,315 @@
+}
+```
+
+### Abstract process
+
+#### Process for certificate issuance
+
+Adopting the `CertificateIssuer` interface fundamentally changes _how_ the CO issues certificates. 
+Because of the current assumptions, the CO generates cluster CA certificates (if they're missing or expired) within a reconciliation, before it knows whether any EE certificate issuance is needed in that reconciliation. Then if it needs to issue EE certificates it just uses the CA certificates that must exist by that point.
+
+The steps required by the `CertificateIssuer` contract are more complicated, and will be described over the following sections. 
+The term "issuing operator" will be used to refer to the operator that's issuing the certificate. This could be the User or Cluster Operator.
+
+Before describing the process let's introduce two state machines. First the "EE certificate state machine"
+
+```
+  NEW // a certificate is needed for the end entity, but doesn't exist
+   |
+   v
+  REQUESTED // a certificate has been requested
+   |
+   v
+  PENDING // a certificate has been issued, but the root CA is not trusted yet
+   |
+   v
+  IN_USE // the certificate is in use
+   |
+   v
+  NOT_NEEDED // a certificate for the end entity is not needed any more
+```
+
+Then the "CA certificate trust state machine":
+
+```
+  UNTRUSTED // not yet trusted everywhere it needs to be 
+   |
+   v
+  TRUSTED_NEW // trusted everywhere it needs to be, but no EE certificates yet issued
+   |
+   v
+  TRUSTED_USED // trusted everywhere, EE certificates issued
+   |
+   v
+  PHASE_OUT // trusted everywhere, but being phased out
+   |
+   v
+  ZOMBIE // still trusted by some component, but not relied on by any party.
+```
+
+Now let's describe the processes.
+
+1. The issuing operator needs to figure out which EE certificates are missing (or in their renewal period, or otherwise need replacing, e.g. via annotation). This need for a certificate might exist independently of the end entity. For example, if we're going to scale up the Kafka cluster then we need a certificate for the new broker before the new broker exists. **The issuing operator puts the end entity in the `NEW` state.**
+2. (**See step 1 in the diagram.**) The issuing operator requests a new certificate. **The issuing operator puts the end entity in the `REQUESTED` state.**
+3. (**See step 2 in the diagram.**) The issuing operator cannot assume that issuance will happen quickly (i.e. that `pollForRequestedCertificate()` returns a non-empty `Optional` immediately), so it needs to be able persist the token obtained from `requestCertificate()` in a `Secret`, _and possibly complete the current reconciliation_ (i.e. issuance can be very asynchronous). 
+(An example of how this could happen is where a `CertificateIssuer` implementation uses cert-manager; `requestCertificate()` creates a cert-manager `CertificateRequest` CR but while the cert manager operator is not running no certificate will be issued).
```

Comment:
You can use `Certificate`, in which case cert manager will keep a Secret up to date for you. I don't think that really helps us though, because we still need to orchestrate the roll out of the EE certificate (and possible the roll out of a new CA cert too). So we'd still need similar machinery in the CO, it's just we'd be needing to spot changes in the `Secret` which cert manager updates spontaneously, rather than being in control of renewal ourselves (and thus being able to respect the maintenance times, for example).

Reply:
> You can use Certificate, in which case cert manager will keep a Secret up to date for you.

Mhhh ... but by using an external certificates manager, so cert-manager for example, aren't we leveraging it to generate certificates and storing them in Secrets? At this point it's not clear to me how we are interacting with a cert-manager, how we get the generated certificates to put them in our own Secrets.

Reply:
The difference, as I understand it, is that with `Certificate` cert manager will renew it automatically (basically creating a `CertificateRequest` behind the scenes) and update the Secret. Whereas `CertificateRequest` is a one time deal, no renewal on expiry.

Reply:
@ppatierno did that clarify your question?

Reply:
so aren't we leveraging to the external cert manager to do the renewal process for us automatically?  we want to control the cert manager to do the renewal when we want to, is that right? And it means interacting with cert-manager, for example, at lower level (`CertificateRequest` vs `Certificate`)?